internal/replacements/removals.go

```diff
@@ -0,0 +1,104 @@
+/*
+ * removals.go
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package replacements
+
+import (
+	"fmt"
+	"time"
+
+	fdbtypes "github.com/FoundationDB/fdb-kubernetes-operator/api/v1beta1"
+	"github.com/FoundationDB/fdb-kubernetes-operator/internal"
+	"github.com/FoundationDB/fdb-kubernetes-operator/pkg/fdbadminclient"
+	"github.com/go-logr/logr"
+)
+
+func getMaxReplacements(cluster *fdbtypes.FoundationDBCluster, maxReplacements int) int {
+	// The maximum number of replacements will be the defined number in the cluster spec
+	// minus all currently ongoing replacements e.g. process groups marked for removal but
+	// not fully excluded.
+	removalCount := 0
+	for _, processGroupStatus := range cluster.Status.ProcessGroups {
+		if processGroupStatus.Remove && (!processGroupStatus.Excluded || processGroupStatus.ExclusionSkipped) {
+			// If we already have a removal in-flight, we should not try
+			// replacing more failed process groups.
+			removalCount++
+		}
+	}
+
+	return maxReplacements - removalCount
+}
+
+// ChooseNewRemovals flags failed processes groups for removal and returns an indicator
+// of whether any processes were thus flagged.
+func ChooseNewRemovals(log logr.Logger, cluster *fdbtypes.FoundationDBCluster, adminClient fdbadminclient.AdminClient) bool {
```

Comment:
It looks like this function is doing a different kind of selection than the ChooseRemovals action. Is that right? If so, we might want to rename one of them to avoid confusion.

Reply:
I'm also not clear on how we're using the term "replacement" and "removal" in these new functions. It seems like `ChooseNewRemovals` and `HasReplacements` are doing the same kind of thing, but for different root problems. It also seems odd to have one labelled `ChooseX` and the other `HasX` if both functions are marking things for removal.